packages/tab/index.tsx

```diff
@@ -0,0 +1,241 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import * as React from 'react';
+import * as classnames from 'classnames';
+import TabIndicator from '@material/react-tab-indicator';
+// No mdc .d.ts files
+// @ts-ignore
+import {MDCTabFoundation} from '@material/tab/dist/mdc.tab';
+import TabRipple, {TabRippleProps} from './TabRipple';
+
+export {
+  TabRipple,
+  Tab,
+  TabRippleProps,
+};
+
+export interface TabProps extends React.HTMLProps<HTMLButtonElement> {
+  active?: boolean;
+  className?: string;
+  isFadingIndicator?: boolean;
+  indicatorContent?: React.ReactNode;
+  minWidth?: boolean;
+  isMinWidthIndicator?: boolean;
+  stacked?: boolean;
+  previousIndicatorClientRect?: ClientRect;
+}
+
+interface TabAttributes {
+  'aria-selected': boolean;
+  tabIndex?: number;
+}
+
+interface TabState extends TabAttributes {
+  classList: Set<string>;
+  activateIndicator: boolean;
+  previousIndicatorClientRect?: ClientRect;
+}
+
+export default class Tab extends React.Component<TabProps, TabState> {
+  foundation?: MDCTabFoundation;
+  tabElement: React.RefObject<HTMLButtonElement> = React.createRef();
+  tabContentElement: React.RefObject<HTMLSpanElement> = React.createRef();
+  tabIndicator: React.RefObject<TabIndicator> = React.createRef();
+  tabRipple: React.RefObject<TabRipple> = React.createRef();
+
+  static defaultProps: Partial<TabProps> = {
+    active: false,
+    className: '',
+    isFadingIndicator: false,
+    indicatorContent: null,
+    minWidth: false,
+    isMinWidthIndicator: false,
+    stacked: false,
+  };
+
+  state: TabState = {
+    'classList': new Set(),
+    'aria-selected': false,
+    'activateIndicator': false,
+    'previousIndicatorClientRect': this.props.previousIndicatorClientRect,
+  };
+
+  componentDidMount() {
+    this.foundation = new MDCTabFoundation(this.adapter);
+    this.foundation.init();
+    if (this.props.active) {
+      this.foundation.activate();
+    }
+  }
+
+  componentWillUnmount() {
+    this.foundation.destroy();
+  }
+
+  componentDidUpdate(prevProps: TabProps) {
+    if (this.props.active !== prevProps.active) {
+      if (this.props.active) {
+        // If active state is updated through props, previousIndicatorClientRect must also be passed through props
+        this.activate(this.props.previousIndicatorClientRect);
+      } else {
+        this.deactivate();
+      }
+    }
+  }
+
+  get classes() {
+    const {classList} = this.state;
+    const {className, minWidth, stacked} = this.props;
+    return classnames('mdc-tab', Array.from(classList), className, {
+      'mdc-tab--min-width': minWidth,
+      'mdc-tab--stacked': stacked,
+    });
+  }
+
+  get adapter() {
+    return {
+      addClass: (className: string) => {
```

Comment:
If you locally alias these:

```typescript
const {
  state,
  setState,
  tabRef,
  tabContentRef,
} = this;
```

you can make these methods instead of arrow functions.  That would hem closer to the ts-style's recommendations that discourage arrow properties.

Reply:
These actually need to access `this.state` every time the method is called. Creating a variable `state` is a reference to state at that moment in time.